### PR TITLE
JEP371 implementation part 2

### PR DESCRIPTION
**Pull request by hangshao0**
1. Add a new API defineClassImpl1() in classLoader.java.
2. Replace Unsafe.defineAnonymousClass() with the new API to 
define hidden classes.
3. Add J9AccClassHiddenClass, J9AccHiddenClassOptionNestmate and
J9AccHiddenClassOptionStrong.
4. Disable class sharing (and SCC tests) for hidden class for now. 

Hidden classes are treated the same as unsafe anonymous 
classes, except: 
a) Its defining classloader is the loader of its hostclass.
b) It has J9AccClassHiddenClass set.
c) If defined with NESTMATE/STRONG,
J9AccHiddenClassOptionNestmate/J9AccHiddenClassOptionStrong 
will be set.

issue #9328

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

### CODE REVIEW

File: runtime/bcutil/ROMClassBuilder.cpp [hunk 1]
```diff
@@ -1059,17 +1059,17 @@ ROMClassBuilder::finishPrepareAndLaydown(
  *                                  + UNUSED
  *                                 + UNUSED
  *                                + UNUSED
- *                               + UNUSED
+ *                              + J9AccHiddenClassOptionNestmate
```
Comment (gacholio) [inline runtime/bcutil/ROMClassBuilder.cpp:1062]: I would prefer these new flags follow the convention - start with J9AccClass

File: runtime/bcutil/ROMClassSegmentAllocationStrategy.cpp [hunk 2]
```diff
@@ -34,9 +34,10 @@ ROMClassSegmentAllocationStrategy::allocate(UDATA bytesRequired)
 	/* Scan existing segments for one large enough to hold the new ROM class */
 
 	J9MemorySegment* segment = NULL;
-	bool isLoadedByAnonClassLoader = _classLoader == _javaVM->anonClassLoader;
 	/* always make a new segment if its an anonClass */
-	if (!isLoadedByAnonClassLoader) {
+	bool allocNewSegment = (_allocNewSeg || _classLoader == _javaVM->anonClassLoader);
```
Comment (gacholio) [inline runtime/bcutil/ROMClassSegmentAllocationStrategy.cpp:38]: Brackets please.

File: runtime/jcl/common/clsldr.cpp [hunk 3]
```diff
@@ -85,6 +93,66 @@ Java_java_lang_ClassLoader_defineClassImpl(JNIEnv *env, jobject receiver, jstrin
...
+	j9object_t hostClassObject = J9_JNI_UNWRAP_REFERENCE(hostClass);
+	J9Class *hostClazz =  J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, hostClassObject);
+
+	vmFuncs->internalExitVMToJNI(currentThread);
+
+	options |= (J9_FINDCLASS_FLAG_HIDDEN | J9_FINDCLASS_FLAG_UNSAFE | J9_FINDCLASS_FLAG_ANON);
+	if (J9_ARE_ALL_BITS_SET(flags, CLASSOPTION_FLAG_NESTMATE)) {
+		options |= J9_FINDCLASS_FLAG_CLASS_OPTION_NESTMATE;
+	}
+	if (J9_ARE_ALL_BITS_SET(flags, CLASSOPTION_FLAG_STRONG )) {
```
Comment (gacholio) [inline runtime/jcl/common/clsldr.cpp:126]: Extra blank

File: runtime/jcl/common/clsldr.cpp [hunk 4]
```diff
@@ -85,6 +93,66 @@ Java_java_lang_ClassLoader_defineClassImpl(JNIEnv *env, jobject receiver, jstrin
...
+	if (init) {
+		vmFuncs->internalEnterVMFromJNI(currentThread);
+		j9object_t classObject = J9_JNI_UNWRAP_REFERENCE(result);
+		J9Class *j9clazz =  J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, classObject);
+		if (VM_VMHelpers::classRequiresInitialization(currentThread, j9clazz)) {
+			vmFuncs->initializeClass(currentThread, j9clazz);
+		}
+		vmFuncs->internalExitVMToJNI(currentThread);
+	}
+
```
Comment (gacholio) [inline runtime/jcl/common/clsldr.cpp:149]: Extra line

File: runtime/jcl/common/java_lang_Class.cpp [hunk 5]
```diff
@@ -1812,17 +1812,6 @@ Java_java_lang_Class_getNestHostImpl(JNIEnv *env, jobject recv)
 
 	if (NULL == nestHost) {
 		J9Class *clazzToUse = clazz;
-#if JAVA_SPEC_VERSION >= 15
```
Comment (gacholio) [inline runtime/jcl/common/java_lang_Class.cpp:1815]: Why is this code being removed?

File: runtime/jcl/common/java_lang_Class.cpp [hunk 6]
```diff
@@ -1937,4 +1926,22 @@ Java_java_lang_Class_getNestMembersImpl(JNIEnv *env, jobject recv)
...
+jboolean JNICALL 
+Java_java_lang_Class_isHiddenImpl(JNIEnv *env, jobject recv)
+{
+#if JAVA_SPEC_VERSION >= 15
+	jboolean result = JNI_FALSE;
+	J9VMThread *currentThread = (J9VMThread*)env;
+	J9InternalVMFunctions *vmFuncs = currentThread->javaVM->internalVMFunctions;
+	vmFuncs->internalEnterVMFromJNI(currentThread);
+	J9Class *clazz = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, J9_JNI_UNWRAP_REFERENCE(recv));
+	result = J9_ARE_ALL_BITS_SET(clazz->romClass->extraModifiers, J9AccClassHiddenClass);
```
Comment (gacholio) [inline runtime/jcl/common/java_lang_Class.cpp:1938]: This should be a macro in j9modifiers_api.h

File: runtime/jcl/common/jcldefine.c [hunk 7]
```diff
@@ -131,14 +131,17 @@ defineClassCommon(JNIEnv *env, jobject classLoaderObject,
...
 	omrthread_monitor_enter(vm->classTableMutex);
-
-	if (vmFuncs->hashClassTableAt(classLoader, utf8Name, utf8Length) != NULL) {
-		/* Bad, we have already defined this class - fail */
-		omrthread_monitor_exit(vm->classTableMutex);
-		if (J9_ARE_NO_BITS_SET(options, J9_FINDCLASS_FLAG_NAME_IS_INVALID)) {
-			vmFuncs->setCurrentException(currentThread, J9VMCONSTANTPOOL_JAVALANGLINKAGEERROR, (UDATA *)*(j9object_t*)className);
+	
+	if (J9_ARE_NO_BITS_SET(options, J9_FINDCLASS_FLAG_HIDDEN)) {
+		/* Hidden class is never added into the hash table */
```
Comment (gacholio) [inline runtime/jcl/common/jcldefine.c:136]: Please move this comment above the if.

File: runtime/jcl/common/jcldefine.c [hunk 8]
```diff
@@ -131,14 +131,17 @@ defineClassCommon(JNIEnv *env, jobject classLoaderObject,
...
+		if (vmFuncs->hashClassTableAt(classLoader, utf8Name, utf8Length) != NULL) {
```
Comment (gacholio) [inline runtime/jcl/common/jcldefine.c:137]: NULL !=

File: runtime/oti/j9.h [hunk 9]
```diff
@@ -124,7 +124,7 @@ typedef struct J9ClassLoaderWalkState {
 #define J9CLASS_IS_MIXED(ramClass) (((J9CLASS_FLAGS(ramClass) >> J9AccClassRAMShapeShift) & OBJECT_HEADER_SHAPE_MASK) == OBJECT_HEADER_SHAPE_MIXED)
 
 #define J9CLASS_IS_EXEMPT_FROM_VALIDATION(clazz) \
-	(J9ROMCLASS_IS_UNSAFE((clazz)->romClass) || J9_ARE_ANY_BITS_SET((clazz)->classFlags, J9ClassIsExemptFromValidation))
+	((J9ROMCLASS_IS_UNSAFE((clazz)->romClass) && !J9ROMCLASS_IS_HIDDEN((clazz)->romClass)) || (J9_ARE_ANY_BITS_SET((clazz)->classFlags, J9ClassIsExemptFromValidation)))
```
Comment (gacholio) [inline runtime/oti/j9.h:127]: Has this case been tested? It took a while to get this right the first time.

File: runtime/oti/j9javaaccessflags.h [hunk 10]
```diff
@@ -118,6 +119,8 @@
 #define J9AccVolatile 0x40
 #define J9AccRecord 0x400
 #define J9AccSealed 0x200
+#define J9AccHiddenClassOptionNestmate 0x80
```
Comment (gacholio) [inline runtime/oti/j9javaaccessflags.h:122]: Are you sure these bits are available? I see overlap with varags (likely doesn't apply to classes) and ValueType.

File: runtime/util/hshelp.c [hunk 11]
```diff
@@ -2469,8 +2469,15 @@ recreateRAMClasses(J9VMThread * currentThread, J9HashTable * classHashTable, J9H
 		if (fastHCR) {
 			options |= J9_FINDCLASS_FLAG_FAST_HCR;
 		}
-
-		if (J9_ARE_ALL_BITS_SET(originalRAMClass->classFlags, J9ClassIsAnonymous)) {
+		if (J9_ARE_ANY_BITS_SET(originalRAMClass->romClass->extraModifiers, J9AccClassHiddenClass)) {
```
Comment (gacholio) [inline runtime/util/hshelp.c:2472]: macro please

File: runtime/util/hshelp.c [hunk 12]
```diff
@@ -3440,7 +3447,16 @@ reloadROMClasses(J9VMThread * currentThread, jint class_count, const jvmtiClassD
 			options = options | J9_FINDCLASS_FLAG_UNSAFE;
 		}
 		loadData.classLoader = originalRAMClass->classLoader;
-		if (J9_ARE_ALL_BITS_SET(originalRAMClass->classFlags, J9ClassIsAnonymous)) {
+		if (J9_ARE_ANY_BITS_SET(originalRAMClass->romClass->extraModifiers, J9AccClassHiddenClass)) {
```
Comment (gacholio) [inline runtime/util/hshelp.c:3450]: macro please

File: runtime/util/hshelp.c [hunk 13]
```diff
@@ -2469,8 +2469,15 @@ recreateRAMClasses(J9VMThread * currentThread, J9HashTable * classHashTable, J9H
 		if (fastHCR) {
 			options |= J9_FINDCLASS_FLAG_FAST_HCR;
 		}
-
-		if (J9_ARE_ALL_BITS_SET(originalRAMClass->classFlags, J9ClassIsAnonymous)) {
+		if (J9_ARE_ANY_BITS_SET(originalRAMClass->romClass->extraModifiers, J9AccClassHiddenClass)) {
+			options |= (J9_FINDCLASS_FLAG_HIDDEN | J9_FINDCLASS_FLAG_UNSAFE | J9_FINDCLASS_FLAG_ANON);
+			if (J9_ARE_ANY_BITS_SET(originalRAMClass->romClass->extraModifiers, J9AccHiddenClassOptionNestmate)) {
```
Comment (gacholio) [inline runtime/util/hshelp.c:2474]: These tests could also be macros in j9modifiers_api.h

File: runtime/vm/ClassInitialization.cpp [hunk 14]
```diff
@@ -157,11 +157,14 @@ performVerification(J9VMThread *currentThread, J9Class *clazz)
 		/* See if this class should be verified:
 		 *
 		 * - Do not verify any class created by sun.misc.Unsafe
+		 * - Do not verify hidden classes
 		 * - Do not verify any class which is marked for exclusion in the optional flags
 		 * - Verify every class whose bytecodes have been modified
 		 * - Do not verify bootstrap classes if the appropriate runtime flag is set
 		 */
-		if (!J9CLASS_IS_EXEMPT_FROM_VALIDATION(clazz) && J9_ARE_NO_BITS_SET(romClass->optionalFlags, J9_ROMCLASS_OPTINFO_VERIFY_EXCLUDE)) {
+		if (((!J9CLASS_IS_EXEMPT_FROM_VALIDATION(clazz)) && J9_ARE_NO_BITS_SET(romClass->extraModifiers, J9AccClassHiddenClass))
```
Comment (gacholio) [inline runtime/vm/ClassInitialization.cpp:165]: Why does the EXEMPT macro allow hidden classes only to have them excluded here? Seems confusing.

File: runtime/vm/classsupport.c [hunk 15]
```diff
@@ -223,7 +223,15 @@ internalCreateArrayClass(J9VMThread* vmThread, J9ROMArrayClass* romClass, J9Clas
 #endif /* defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
 
 	if (elementInitSuccess) {
-		if (J9_ARE_ANY_BITS_SET(elementClass->classFlags, J9ClassIsAnonymous)) {
+		if (J9_ARE_ANY_BITS_SET(elementClass->romClass->extraModifiers, J9AccClassHiddenClass)) {
```
Comment (gacholio) [inline runtime/vm/classsupport.c:226]: macros please

File: runtime/vm/visible.c [hunk 16]
```diff
@@ -283,9 +283,20 @@ loadAndVerifyNestHost(J9VMThread *vmThread, J9Class *clazz, UDATA options)
 		J9UTF8 *nestHostName = J9ROMCLASS_NESTHOSTNAME(romClass);
 		BOOLEAN canRunJavaCode = J9_ARE_NO_BITS_SET(options, J9_LOOK_NO_JAVA);
 		BOOLEAN throwException = canRunJavaCode && J9_ARE_NO_BITS_SET(options, J9_LOOK_NO_THROW);
+		BOOLEAN hiddenNestMate = J9_ARE_ALL_BITS_SET(romClass->extraModifiers, J9AccHiddenClassOptionNestmate);
+
+		J9Class* curClazz = clazz;
+		BOOLEAN isCurClassHiddenNestMate = hiddenNestMate;
+		while (isCurClassHiddenNestMate && curClazz != curClazz->hostClass) {
```
Comment (gacholio) [inline runtime/vm/visible.c:290]: brackets please

**Comment by gacholio** [comment]
Code seems correct at a glance (given I have no knowledge of the JEP). I glanced at the java code and didn't look at the test stuff at all.

File: runtime/bcutil/ROMClassBuilder.cpp (hunk 1 above)
Comment (hangshao0) [reply runtime/bcutil/ROMClassBuilder.cpp]: Fixed.

File: runtime/bcutil/ROMClassSegmentAllocationStrategy.cpp (hunk 2 above)
Comment (hangshao0) [reply runtime/bcutil/ROMClassSegmentAllocationStrategy.cpp]: Fixed.

File: runtime/jcl/common/clsldr.cpp (hunk 3 above)
Comment (hangshao0) [reply runtime/jcl/common/clsldr.cpp]: Fixed.

File: runtime/jcl/common/clsldr.cpp (hunk 4 above)
Comment (hangshao0) [reply runtime/jcl/common/clsldr.cpp:149]: Removed.

File: runtime/jcl/common/java_lang_Class.cpp (hunk 5 above)
Comment (hangshao0) [reply runtime/jcl/common/java_lang_Class.cpp:1815]: Our previous implementation used Unsafe.defineAnonymousClass() to define hidden classes. It did not take ClassOption.NESTMATE that is passed in, so we returned the nest host of hostClass for anonymous classes here as a temporary workaround to make java -version working. Change in this PR is taking care of the ClassOption.NESTMATE and setting J9AccHiddenClassOptionNestmate. The change in loadAndVerifyNestHost() in this PR is now checking J9AccHiddenClassOptionNestmate and returning the correct nest host for hidden classes, so this piece of code can be removed.

File: runtime/jcl/common/java_lang_Class.cpp (hunk 6 above)
Comment (hangshao0) [reply runtime/jcl/common/java_lang_Class.cpp]: Fixed.

File: runtime/jcl/common/jcldefine.c (hunk 7 above)
Comment (hangshao0) [reply runtime/jcl/common/jcldefine.c]: Fixed.

File: runtime/oti/j9.h (hunk 9 above)
Comment (hangshao0) [reply runtime/oti/j9.h:127]: Yes, it is tested. This does not change behavior for all unsafe non-hidden classes.

File: runtime/util/hshelp.c (hunk 11 above)
Comment (hangshao0) [reply runtime/util/hshelp.c]: Fixed.

File: runtime/util/hshelp.c (hunk 13 above)
Comment (hangshao0) [reply runtime/util/hshelp.c]: Added macros.

File: runtime/util/hshelp.c (hunk 12 above)
Comment (hangshao0) [reply runtime/util/hshelp.c]: Fixed.

File: runtime/vm/classsupport.c (hunk 15 above)
Comment (hangshao0) [reply runtime/vm/classsupport.c]: Fixed.

File: runtime/vm/visible.c (hunk 16 above)
Comment (hangshao0) [reply runtime/vm/visible.c]: Fixed.

File: runtime/vm/ClassInitialization.cpp (hunk 14 above)
Comment (hangshao0) [reply runtime/vm/ClassInitialization.cpp]: If I allow J9AccClassHiddenClass go through this, we will eventually go into https://github.com/eclipse/openj9/blob/master/runtime/bcverify/vrfyhelp.c#L957. As hidden class is not in the hashtable, we won't be find the ramclass, causing crash later inside the verifier.

File: runtime/jcl/common/jcldefine.c (hunk 8 above)
Comment (hangshao0) [reply runtime/jcl/common/jcldefine.c]: Fixed.

File: runtime/oti/j9javaaccessflags.h (hunk 10 above)
Comment (hangshao0) [reply runtime/oti/j9javaaccessflags.h]: > However, I noticed both J9AccVarArgs and J9AccTransient are 0x80, and they are both set to romClass->modifiers.

This comment can be ignored. After a closer look, they are setting to romMethod->modifiers and romField->modifiers.

File: runtime/bcutil/ROMClassBuilder.cpp [hunk 17]
```diff
@@ -1059,17 +1059,17 @@ ROMClassBuilder::finishPrepareAndLaydown(
...
+ *                             + J9AccClassHiddenOptionStrong
  *                            + AccSealed
  *                           + AccRecord
  *                          + AccClassAnonClass
  *
  *                        + AccSynthetic (matches Oracle modifier position)
  *                       + AccClassUseBisectionSearch
  *                      + AccClassInnerClass
- *                     + UNUSED
+ *                     + J9AccClassHiddenClass
```
Comment (gacholio) [inline runtime/bcutil/ROMClassBuilder.cpp:1072]: The trailing `Class` here seems superfluous.

**Comment by gacholio** [comment]
Please tell me what hidden classes are - are they magic defined by the JCL, or do they come from user class files?

I'm concerned about the lack of verification - either the class needs to follow the rules or it is implicitly trusted, you can't have it both ways.

**Comment by hangshao0** [comment]
> Please tell me what hidden classes are - are they magic defined by the JCL, or do they come from user class files?

They are not from class files. They are magic defined by the JCL. JCL defines a hidden class by directly passing in class data bytes to defineHiddenClass():
https://github.com/eclipse/openj9/blob/master/jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java#L2298

Unsafe.defineAnonymousClass() is deprecated and is going to be removed in the future. Hidden class is introduced by this JEP to replace unsafe AnonymousClass. This PR treats hidden class the same as unsafe AnonymousClass except a), b), c) mentioned in the description and the J9CLASS_IS_EXEMPT_FROM_VALIDATION marco. I exclude hidden classes from J9CLASS_IS_EXEMPT_FROM_VALIDATION so that we do some more validation for hidden classes than unsafe AnonymousClass, which is required by some hidden class tests I saw.

**Comment by gacholio** [comment]
If they are generated by the JCL, then they should simply be exempt from validation, which will also fix your verifier issue.

**Comment by hangshao0** [comment]
Another thing is that user can optionally pass in ClassOption.NESTMATE, ClassOption.STRONG to defineHiddenClass().
NESTMATE means the hidden class will be added to the same nest of its lookup class, i.e. the class that calls defineHiddenClass().
STRONG means that the hidden class will be unloaded when its class loader is GC'ed. If STRONG is not used, hidden class can be unloaded earlier once it is unreachable.

NESTMATE is handled by this PR. STRONG is not handled by this PR for now, only its flag is added. I will handle it in the next PR.

**Comment by gacholio** [comment]
Neither of those points contradicts what I'm suggesting.

**Comment by hangshao0** [comment]
> If they are generated by the JCL, then they should simply be exempt from validation, which will also fix your verifier issue.

Yes, I exempted hidden classes in performVerification() by adding !J9ROMCLASS_IS_HIDDEN(romClass) after the J9CLASS_IS_EXEMPT_FROM_VALIDATION() check. Hidden classes won't goes into j9bcv_verifyBytecodes().

The reason I change J9CLASS_IS_EXEMPT_FROM_VALIDATION is that some tests require us to do access/visibility checks in processMethod(). It expects us to throw illegal access error if a hidden class is calling, for example, a private method of another class, I changed J9CLASS_IS_EXEMPT_FROM_VALIDATION so that we can go into the access/visibility checks.

File: runtime/bcutil/ROMClassBuilder.cpp (hunk 17 above)
Comment (gacholio) [reply runtime/bcutil/ROMClassBuilder.cpp:1072]: I'd still like this changed.

Comment (hangshao0) [reply runtime/bcutil/ROMClassBuilder.cpp:1072]: > I'd still like this changed.

Yes, I will change this. But after lunch.

Comment (hangshao0) [reply runtime/bcutil/ROMClassBuilder.cpp]: Fixed.

**Comment by gacholio** [comment]
jenkins test sanity,extended win jdk15

**Comment by gacholio** [comment]
jenkins test sanity,extended zlinux jdk8

**Comment by hangshao0** [comment]
Looking at J9CLASS_IS_EXEMPT_FROM_VALIDATION, I am curious why exempt unsafe classes from validation ?

**Comment by gacholio** [comment]
Because they used to be used for reflect accessors, which generate code that won't work if the visibility rules are followed.

**Comment by gacholio** [comment]
Compile errors, please have a look.

**Comment by hangshao0** [comment]
Fixes the compile error. 

I mean we still exempt calling into j9bcv_verifyBytecodes() for unsafe classes.

**Comment by gacholio** [comment]
jenkins test sanity,extended win jdk15

**Comment by gacholio** [comment]
> I mean we still exempt calling into j9bcv_verifyBytecodes() for unsafe classes.

Yes - today various unsafe defined classes will absolutely not verify.

**Comment by gacholio** [comment]
jenkins test sanity,extended zlinux jdk8

**Pull request merged**